File: reader_client/src/lib/api/books.ts
/**
 * Books API
 *
 * Endpoints for book listing, details, and chapter data.
 */

import { api, type ApiRequestOptions } from './client';
import type {
	Book,
	BookDetail,
	BooksResponse,
	ChapterMeta,
	ChapterPassagesResponse,
	ExpandMode,
	PassageBase,
	PassageWithAnnotations,
	PassageFull,
	Testament
} from './types';

// ─────────────────────────────────────────────────────────────────────────────
// Book Listing
// ─────────────────────────────────────────────────────────────────────────────

/**
 * List all books, optionally filtered by testament
 */
export async function listBooks(
	testament?: Testament,
	options?: ApiRequestOptions
): Promise<BooksResponse> {
	return api.get<BooksResponse>('/books', { testament }, options);
}

/**
 * Get detailed book info including chapter list
 */
export async function getBook(bookId: string): Promise<BookDetail> {
	return api.get<BookDetail>(`/books/${bookId}`);
}

// ─────────────────────────────────────────────────────────────────────────────
// Chapter Data
// ─────────────────────────────────────────────────────────────────────────────

/**
 * Get chapter metadata (verse count, first/last verse IDs)
 */
export async function getChapterMeta(
	bookId: string,
	chapter: number
): Promise<ChapterMeta> {
	return api.get<ChapterMeta>(`/books/${bookId}/chapters/${chapter}`);
}

/**
 * Get chapter passages - the primary reader endpoint
 *
 * @param expand - 'none' for fast scanning, 'annotations' for reader display,
 *                 'full' for complete data including HTML
 */
export async function getChapterPassages(
	bookId: string,
	chapter: number,
	expand: 'none',
	options?: ApiRequestOptions
): Promise<ChapterPassagesResponse<PassageBase>>;

export async function getChapterPassages(
	bookId: string,
	chapter: number,
	expand: 'annotations',
	options?: ApiRequestOptions
): Promise<ChapterPassagesResponse<PassageWithAnnotations>>;

export async function getChapterPassages(
	bookId: string,
	chapter: number,
	expand: 'full',
	options?: ApiRequestOptions
): Promise<ChapterPassagesResponse<PassageFull>>;

export async function getChapterPassages(
	bookId: string,
	chapter: number,
	expand?: ExpandMode,
	options?: ApiRequestOptions
): Promise<ChapterPassagesResponse>;

export async function getChapterPassages(
	bookId: string,
	chapter: number,
	expand: ExpandMode = 'annotations',
	options?: ApiRequestOptions
): Promise<ChapterPassagesResponse> {
	return api.get<ChapterPassagesResponse>(
		`/books/${bookId}/chapters/${chapter}/passages`,
		{ expand },
		options
	);
}

/**
 * Get a range of verses within a chapter
 */
export async function getVerseRange(
	bookId: string,
	chapter: number,
	verseStart: number,
	verseEnd: number,
	expand: ExpandMode = 'annotations'
): Promise<ChapterPassagesResponse> {
	return api.get<ChapterPassagesResponse>(
		`/books/${bookId}/chapters/${chapter}/passages`,
		{
			expand,
			verse_start: verseStart,
			verse_end: verseEnd
		}
	);
}

// ─────────────────────────────────────────────────────────────────────────────
// Convenience Exports
// ─────────────────────────────────────────────────────────────────────────────

export const books = {
	list: listBooks,
	get: getBook,
	getChapterMeta,
	getChapterPassages,
	getVerseRange
};

File: api/routers/books.py
"""Book endpoints."""

from fastapi import APIRouter, HTTPException

from api.models.book import BookDetail, BookListResponse, ChapterDetail
from api.models.common import ExpandMode, Testament
from api.models.passage import (
    ChapterPassagesResponse,
    PassageFull,
    PassageMinimal,
    PassageWithAnnotations,
)
from api.services import book_service, passage_service

router = APIRouter(prefix="/books", tags=["books"])


@router.get("", response_model=BookListResponse)
async def list_books(testament: Testament | None = None):
    """List all books, optionally filtered by testament."""
    return await book_service.get_books(testament)


@router.get("/{book_id}", response_model=BookDetail)
async def get_book(book_id: str):
    """Get book details with chapter breakdown."""
    book = await book_service.get_book(book_id)
    if not book:
        raise HTTPException(status_code=404, detail=f"Book not found: {book_id}")
    return book


@router.get("/{book_id}/chapters/{chapter}", response_model=ChapterDetail)
async def get_chapter(book_id: str, chapter: int):
    """Get chapter metadata."""
    chapter_detail = await book_service.get_chapter(book_id, chapter)
    if not chapter_detail:
        raise HTTPException(
            status_code=404,
            detail=f"Chapter not found: {book_id} chapter {chapter}",
        )
    return chapter_detail


@router.get("/{book_id}/chapters/{chapter}/passages")
async def get_chapter_passages(
    book_id: str,
    chapter: int,
    expand: ExpandMode = ExpandMode.NONE,
    verse_start: int | None = None,
    verse_end: int | None = None,
) -> ChapterPassagesResponse:
    """
    Get all passages in a chapter.

    Optimized for reader component - full chapter fetch with optional annotations.
    Use `verse_start` and `verse_end` to filter to a verse range.

    Includes chapter navigation with cross-book support (Genesis 50 -> Exodus 1).
    """
    # Verify book exists
    book = await book_service.get_book(book_id)
    if not book:
        raise HTTPException(status_code=404, detail=f"Book not found: {book_id}")

    passages, navigation = await passage_service.get_chapter_passages(
        book_id, chapter, expand, verse_start, verse_end
    )

    if not passages:
        raise HTTPException(
            status_code=404,
            detail=f"No passages found: {book_id} chapter {chapter}",
        )

    return ChapterPassagesResponse(
        book_id=book_id,
        book_name=book.name,
        chapter=chapter,
        passages=passages,
        total=len(passages),
        navigation=navigation,
    )
